tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/ServiceCommand.java

```diff
@@ -0,0 +1,294 @@
+
+package com.liferay.ide.project.core.modules;
+
+import java.io.InputStream;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+import aQute.remote.api.Agent;
+import aQute.remote.api.Event;
+import aQute.remote.api.Supervisor;
+import aQute.remote.util.AgentSupervisor;
+
+/**
+ * @author Lovett Li
+ */
+public class ServiceCommand {
+
+    private String _serviceName;
+
+    public ServiceCommand()
+    {
+    }
+    public ServiceCommand( String serviceName )
+    {
+        _serviceName = serviceName;
+    }
+
+    public String[] execute() throws Exception
+    {
+        ServicesSupervisor supervisor = null;
+        String[] result = null;
+
+        synchronized( ServiceCommand.class )
+        {
+
+            try
+            {
+                supervisor = connectRemote();
+
+                if( _serviceName == null )
+                {
+                    result = getServices( supervisor );
+                }
+                else
+                {
+                    result = getServiceBundle( _serviceName, supervisor );
+                }
+
+                return result;
+            }
+            finally
+            {
+                if( supervisor != null )
+                {
+                    supervisor.close();
+                }
+
+            }
+        }
+    }
+
+    private ServicesSupervisor connectRemote() throws Exception
+    {
+        if( !canConnect( "localhost", Agent.DEFAULT_PORT ) )
+        {
+            return null;
+        }
+
+        ServicesSupervisor supervisor = new ServicesSupervisor();
+        supervisor.connect( "localhost", Agent.DEFAULT_PORT );
+
+        if( !supervisor.getAgent().redirect( -1 ) )
+        {
+            supervisor.close();
+            return null;
+        }
+
+        return supervisor;
+    }
+
+    static boolean canConnect( String host, int port )
+    {
+        InetSocketAddress address = new InetSocketAddress( host, Integer.valueOf( port ) );
+        InetSocketAddress local = new InetSocketAddress( 0 );
+
+        InputStream in = null;
+
+        try(Socket socket = new Socket())
+        {
+            socket.bind( local );
+            socket.connect( address, 3000 );
+            in = socket.getInputStream();
+
+            return true;
+        }
+        catch( Exception e )
+        {
+        }
+        finally
+        {
+            if( in != null )
+            {
+                try
+                {
+                    in.close();
+                }
+                catch( Exception e )
+                {
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private String[] getServices( ServicesSupervisor supervisor ) throws Exception
+    {
+        supervisor.getAgent().stdin( "services" );
```

Comment:
always add a newlien before return unless return is the only statement in the block.